indexing-service/src/main/java/org/apache/druid/indexing/common/task/AbstractBatchIndexTask.java

```diff
@@ -576,6 +582,73 @@ static Granularity findGranularityFromSegments(List<DataSegment> segments)
     }
   }
 
+  /**
+   * Wait for segments to become available on the cluster. If waitTimeout is reached, giveup on waiting. This is a
+   * QoS method that can be used to make Batch Ingest tasks wait to finish until their ingested data is available on
+   * the cluster. Doing so gives an end user assurance that a Successful task status means their data is available
+   * for querying.
+   *
+   * @param toolbox {@link TaskToolbox} object with for assisting with task work.
+   * @param segmentsToWaitFor {@link List} of segments to wait for availability.
+   * @param waitTimeout Millis to wait before giving up
+   * @return True if all segments became available, otherwise False.
+   */
+  protected boolean waitForSegmentAvailability(TaskToolbox toolbox, ExecutorService exec, List<DataSegment> segmentsToWaitFor, long waitTimeout)
+  {
+    if (segmentsToWaitFor.isEmpty()) {
+      log.info("Asked to wait for segments to be available, but I wasn't provided with any segments!?");
+      return false;
+    }
+    log.info("Waiting for segments to be loaded by the cluster...");
+
+    SegmentHandoffNotifier notifier = toolbox.getSegmentHandoffNotifierFactory()
+                                             .createSegmentHandoffNotifier(segmentsToWaitFor.get(0).getDataSource());
+
+    notifier.start();
+    for (DataSegment s : segmentsToWaitFor) {
+      notifier.registerSegmentHandoffCallback(
+          new SegmentDescriptor(s.getInterval(), s.getVersion(), s.getShardSpec().getPartitionNum()),
+          exec,
+          () -> {
+            log.info(
+                "Confirmed availability for [%s]. Removing from list of segments to wait for",
+                s.getId()
+            );
+            synchronized (availabilityCondition) {
+              segmentsToWaitFor.remove(s);
```

Comment:
not sure but removing items from the collection while iterating it may cause an error. Instead, we could use a counter and a completableFuture together. Something like this 
```
CompletableFuture<Void> uberFuture = CompletableFuture.completedFuture(null);
for (DataSegment s : segmentsToWaitFor) {
      CompletableFuture<Void> future = new CompletableFuture<>();
      notifier.registerSegmentHandoffCallback(
          new SegmentDescriptor(s.getInterval(), s.getVersion(), s.getShardSpec().getPartitionNum()),
          exec,
          () -> {
            log.info(
                "Confirmed availability for [%s]. Removing from list of segments to wait for",
                s.getId()
            );
           future.complete(null);
         }
      );
     uberFuture = uberFuture.thenCombine(future, (a, b) -> null);
    }
uberFuture.get(waitTimeout, TimeUnit.MILLISECONDS)
```

Reply:
yes, I've been trying to come up with an alternative to avoid concurrent access risks. thanks so much for this idea

Reply:
I ended up trying a countdown latch with timeout since I am more familiar with those from use in the past as compared to your suggestion. let me know what you think of it

Reply:
Looks good to me @capistrant. Thanks.